### PR TITLE
chore: more tests for catalog work

### PR DESCRIPTION
**Pull request by jhaals**
Signed-off-by: Johan Haals <johan.haals@gmail.com>

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr5542-w9gixcivafucvp7knisktrg8gol6pgos.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/608bb812a55c9aff66a59ac0

File: plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.ts [hunk 1]
```diff
@@ -87,6 +87,7 @@ export class DefaultProcessingDatabase implements ProcessingDatabase {
       .update({
         processed_entity: JSON.stringify(processedEntity),
         cache: JSON.stringify(state),
+        // TODO: should this also be JSON stringfy?
```
Comment (freben) [inline plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.ts:90]: no, seems to be a string in the type already

File: plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.test.ts [hunk 2]
```diff
@@ -23,33 +23,187 @@ import {
...
+          deferredEntities: [],
+          errors: 'something broke',
+        }),
       );
-    };
 
-    const insertRefreshStateRow = async (ref: DbRefreshStateRow) => {
-      await db<DbRefreshStateRow>('refresh_state').insert(ref);
-    };
+      const [entity] = await db<DbRefreshStateRow>('refresh_state').select();
```
Comment (freben) [inline plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.test.ts:111]: would be better to do `const entities =` and then `expect(entities.length).toBe(1)` etc

File: plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.test.ts [hunk 3]
```diff
@@ -23,33 +23,187 @@ import {
...
+      state.set('hello', { t: 'something' });
+
+      await processingDatabase.transaction(tx =>
+        processingDatabase.updateProcessedEntity(tx, {
+          id,
+          processedEntity,
+          state,
+          relations: [],
+          deferredEntities: [],
+          errors: 'something broke',
```
Comment (freben) [inline plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.test.ts:107]: this is defined to be a stringified `Error[]` (so an array); of course the test doesn't care but if you wanted to choose a value here then maybe `'[]'` would be clearest

File: plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.test.ts [hunk 4]
```diff
@@ -23,33 +23,187 @@ import {
...
+
+    it('removes old relations and stores the new relationships', async () => {
+      await insertRefreshStateRow({
+        entity_id: id,
+        entity_ref: 'location:default/fakelocation',
+        unprocessed_entity: '{}',
+        processed_entity: '{}',
+        errors: '',
+        next_update_at: 'now()',
+        last_discovery_at: 'now()',
```
Comment (freben) [inline plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.test.ts:126]: confusing to put the string `now()` here, perhaps pick a real datetime string instead

File: plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.test.ts [hunk 5]
```diff
@@ -23,33 +23,187 @@ import {
...
+        processingDatabase.updateProcessedEntity(tx, {
+          id,
+          processedEntity,
+          state: new Map<string, JsonObject>(),
+          relations: relations,
+          deferredEntities: [],
+        }),
+      );
+
+      const [savedRelations] = await db<DbRelationsRow>('relations')
```
Comment (freben) [inline plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.test.ts:155]: same here, best to get all of them as an array and assert that you got exactly one row back instead

File: plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.test.ts [hunk 6]
```diff
@@ -23,33 +23,187 @@ import {
...
+
+    it('adds deferred entities to the the refresh_state table to be picked up later', async () => {
+      await insertRefreshStateRow({
+        entity_id: id,
+        entity_ref: 'location:default/fakelocation',
+        unprocessed_entity: '{}',
+        processed_entity: '{}',
+        errors: '',
+        next_update_at: 'now()',
+        last_discovery_at: 'now()',
```
Comment (freben) [inline plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.test.ts:175]: same here

File: plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.test.ts [hunk 7]
```diff
@@ -431,86 +585,48 @@ describe('Default Processing Database', () => {
...
-      });
-    });
+  describe('getProcessableEntities', () => {
+    it('should return entities to process', async () => {
+      const entity = JSON.stringify({
+        apiVersion: '1.0.0',
+        metadata: {
+          name: 'xyz',
+        },
+        kind: 'Location',
```
Comment (freben) [inline plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.test.ts:595]: hehe, a  bit weird to see the kind below metadata :)

File: plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.test.ts [hunk 8]
```diff
@@ -431,86 +585,48 @@ describe('Default Processing Database', () => {
...
-        },
-        type: 'url',
-      };
+      await db<DbRefreshStateRow>('refresh_state').insert({
+        entity_id: '1',
+        entity_ref: 'location:default/foobar',
+        unprocessed_entity: entity,
+        errors: '',
+        next_update_at: '2042-01-01 23:00:00',
+        last_discovery_at: 'now()',
```
Comment (freben) [inline plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.test.ts:613]: fake

File: plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.ts (hunk 1 above)
Comment (benjdlambert) [reply plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.ts:90]: But what about multiple errors? Are they just stored as multiline strings? Seems a little strange to do that.

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr5542-w9gixcivafucvp7knisktrg8gol6pgos.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/608bb812a55c9aff66a59ac0

File: plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.ts (hunk 1 above)
Comment (freben) [reply plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.ts:90]: yeah see below, this is expected to be an array of stringified errors currently - but that's not formalized yet, we just left it as a string and then we'll figure out the exact encoding of each array item later

File: plugins/catalog-backend/src/next/DefaultCatalogProcessingEngine.test.ts [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+            },
+            state: new Map(),
+            nextUpdateAt: '',
+            lastDiscoveryAt: '',
+          },
+        ],
+      });
+
+    await engine.start();
+    await new Promise<void>(resolve => setTimeout(resolve, 1));
```
Comment (freben) [inline plugins/catalog-backend/src/next/DefaultCatalogProcessingEngine.test.ts:86]: https://www.npmjs.com/package/wait-for-expect ? :)

File: plugins/catalog-backend/src/next/DefaultCatalogProcessingEngine.ts [hunk 10]
```diff
@@ -16,51 +16,54 @@
...
   ) {}
 
   async applyMutation(mutation: EntityProviderMutation): Promise<void> {
+    const db = this.config.processingDatabase;
     if (mutation.type === 'full') {
-      await this.config.stateManager.replaceProcessingItems({
-        sourceKey: this.config.id,
-        type: 'full',
-        items: mutation.entities,
+      db.transaction(async tx => {
```
Comment (freben) [inline plugins/catalog-backend/src/next/DefaultCatalogProcessingEngine.ts:40]: need to await this one

File: plugins/catalog-backend/src/next/DefaultCatalogProcessingEngine.ts [hunk 11]
```diff
@@ -16,51 +16,54 @@
...
-
       return;
     }
-
-    await this.config.stateManager.replaceProcessingItems({
-      sourceKey: this.config.id,
-      type: 'delta',
-      added: mutation.added,
-      removed: mutation.removed,
+    db.transaction(async tx => {
```
Comment (freben) [inline plugins/catalog-backend/src/next/DefaultCatalogProcessingEngine.ts:49]: await here too

File: plugins/catalog-backend/src/next/DefaultCatalogProcessingEngine.ts [hunk 12]
```diff
@@ -69,51 +72,60 @@ export class DefaultCatalogProcessingEngine implements CatalogProcessingEngine {
...
+
+        const result = await this.orchestrator.process({
+          entity: unprocessedEntity,
+          state,
+        });
+        for (const error of result.errors) {
+          this.logger.warn(error.message);
+        }
+        if (!result.ok) {
+          return;
```
Comment (freben) [inline plugins/catalog-backend/src/next/DefaultCatalogProcessingEngine.ts:103]: should be continue, not return, right? (another reason to go for the wrapping in a separate function as suggested above)

File: plugins/catalog-backend/src/next/DefaultCatalogProcessingEngine.ts [hunk 13]
```diff
@@ -69,51 +72,60 @@ export class DefaultCatalogProcessingEngine implements CatalogProcessingEngine {
...
-      for (const error of result.errors) {
-        this.logger.warn(error.message);
-      }
-
-      if (!result.ok) {
-        return;
+      if (items.length) {
+        const { id, state, unprocessedEntity } = items[0];
+
+        const result = await this.orchestrator.process({
```
Comment (freben) [inline plugins/catalog-backend/src/next/DefaultCatalogProcessingEngine.ts:95]: i'm noting that none of the `await`s in here have a try/catch - if anything fails, the entire processing loop will quit, won't it? we should probably wrap the entire thing in like a `processOneItem` or whatever that's called in a try-catch and which is free to return and throw and whatever without hurting the loop

File: plugins/catalog-backend/src/next/NextCatalogBuilder.ts [hunk 14]
```diff
@@ -252,7 +251,7 @@ export class NextCatalogBuilder {
     const db = new CommonDatabase(dbClient, logger);
 
     const processingDatabase = new DefaultProcessingDatabase(dbClient, logger);
-    const stateManager = new DefaultProcessingStateManager(processingDatabase);
+    // const stateManager = new DefaultProcessingStateManager(processingDatabase);
```
Comment (freben) [inline plugins/catalog-backend/src/next/NextCatalogBuilder.ts:254]: remove?

File: plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.ts (hunk 1 above)
Comment (freben) [reply plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.ts:90]: so probably remove the comment then

**Review by freben — APPROVED** [review]

**Pull request merged**